code/game/machinery/PDApainter.dm

```diff
@@ -1,66 +1,106 @@
+	name = "\improper color manipulator"
+	desc = "A machine able to color PDAs and IDs with ease. Insert an ID card or PDA and pick a color scheme."
 	icon = 'icons/obj/pda.dmi'
-	icon_state = "pdapainter"
+	icon_state = "coloriser"
 	density = TRUE
 	anchored = TRUE
 	var/obj/item/device/pda/storedpda = null
-	var/list/colorlist = list()
+	var/obj/item/card/id/storedid = null
+	var/pda_icons = list(
+		"Assistant" = "pda",
+		"Atmospheric Technician" = "pda-atmos",
+		"Bartender" = "pda-bartender",
+		"Botanist" = "pda-hydro",
+		"Captain" = "pda-captain",
+		"Cargo Technician" = "pda-cargo",
+		"Chaplain" = "pda-chaplain",
+		"Chemist" = "pda-chemistry",
+		"Chief Medical Officer" = "pda-cmo",
+		"Chief Engineer" = "pda-ce",
+		"Clown" = "pda-clown",
+		"Cook" = "pda-cook",
+		"Curator" = "pda-library",
+		"Detective" = "pda-detective",
+		"Engineer" = "pda-engineer",
+		"Geneticist" = "pda-genetics",
+		"Head of Personnel" = "pda-hop",
+		"Head of Security" = "pda-hos",
+		"Internal Affairs Agent" = "pda-laweyr",
+		"Janitor" = "pda-janitor",
+		"Medical Doctor" = "pda-medical",
+		"Mime" = "pda-mime",
+		"Quartermaster" = "pda-qm",
+		"Research Director" = "pda-rd",
+		"Roboticist" = "pda-robotocist",
+		"Scienctist" = "pda-science",
+		"Security Officer" = "pda-security",
+		"Shaft Miner" = "pda-miner",
+		"Virologist" = "pda-virology",
+		"Warden" = "pda-warden")
+	var/id_icons = list(
+		"Assistant" = "id",
+		"Captain" = "gold",
+		"Cargo" = "cargo",
+		"Chief Engineer" = "CE",
+		"Chief Medical Officer" = "CMO",
+		"Clown" = "clown",
+		"Engineering" = "engineering",
+		"Head of Personnel" = "silver",
+		"Head of Security" = "HoS",
+		"Medical" = "medical",
+		"Mime" = "mime",
+		"Research Director" = "RD",
+		"Science" = "research",
+		"Security" = "security")
 	max_integrity = 200
 
-
 /obj/machinery/pdapainter/update_icon()
 	cut_overlays()
 
 	if(stat & BROKEN)
-		icon_state = "[initial(icon_state)]-broken"
+		icon_state = "coloriser-broken"
 		return
 
 	if(storedpda)
-		add_overlay("[initial(icon_state)]-closed")
+		add_overlay("coloriser-pda-in")
+
+	if(storedid)
+		add_overlay("coloriser-id-in")
 
 	if(powered())
 		icon_state = initial(icon_state)
 	else
-		icon_state = "[initial(icon_state)]-off"
+		icon_state = "coloriser-off"
 
 	return
 
-/obj/machinery/pdapainter/Initialize()
-	. = ..()
-	var/list/blocked = list(
-		/obj/item/device/pda/ai/pai,
-		/obj/item/device/pda/ai,
-		/obj/item/device/pda/heads,
-		/obj/item/device/pda/clear,
-		/obj/item/device/pda/syndicate)
-
-	for(var/P in typesof(/obj/item/device/pda) - blocked)
-		var/obj/item/device/pda/D = new P
-
-		//D.name = "PDA Style [colorlist.len+1]" //Gotta set the name, otherwise it all comes up as "PDA"
-		D.name = D.icon_state //PDAs don't have unique names, but using the sprite names works.
-
-		src.colorlist += D
-
 /obj/machinery/pdapainter/Destroy()
 	QDEL_NULL(storedpda)
+	QDEL_NULL(storedid)
 	return ..()
 
 /obj/machinery/pdapainter/on_deconstruction()

```

Comment:
Once you make that single eject proc, use it here so as not to write the same code multiple times.

Reply:
There already are
`ejectid()`
`ejectpda()`